refactor!: signing and the `Signer` trait (#1241)
FuelLabs/fuels-rs: docs/src/custom-transactions/transaction-builders.md

{{#include ../../../examples/cookbook/src/lib.rs:custom_tx_io}}
```

As we have used coins that require a signature, we have to add the signer to the transaction builder with:

```rust,ignore
{{#include ../../../examples/cookbook/src/lib.rs:custom_tx_add_signer}}
```

> **Note** The signature is not created until the transaction is finalized with `build(&provider)`

We need to do one more thing before we stop thinking about transaction inputs. Executing the transaction also incurs a fee that is paid with the base asset. Our base asset inputs need to be large enough so that the total amount covers the transaction fee and any other operations we are doing. The `Account` trait lets us use `adjust_for_fee()` for adjusting the transaction inputs if needed to cover the fee. The second argument to `adjust_for_fee()` is the total amount of the base asset that we expect our transaction to spend regardless of fees. In our case, this is the **ask_amount** we are transferring to the predicate.

```rust,ignore
{{#include ../../../examples/cookbook/src/lib.rs:custom_tx_adjust}}
```

> > **Note** It is recommended to add signers before adjusting for fee as the estimation will include the witnesses fee.

Reviewer: ```suggestion
> **Note** It is recommended to add signers before calling `adjust_for_fee()` as the estimation will include the size of the witnesses.
```

```rust,ignore
{{#include ../../../examples/cookbook/src/lib.rs:custom_tx_verify}}
```

## Building a transaction without signatures

> If you need to build the transaction without signatures - useful when estimating transaction costs or simulations,  you can use the `build_without_signatures(&provider)` method and later sign the built transaction.

Reviewer 2: ```suggestion
If you need to build the transaction without signatures, which is useful when estimating transaction costs or simulations, you can use the `build_without_signatures(&provider)` method and later sign the built transaction.
```

```rust,ignore
{{#include ../../../packages/fuels/tests/contracts.rs:tb_build_without_signatures}}
```

> > **Note** In contrast to signing a transaction builder, when signing a built transaction, you will have to make sure that the order of signatures matches the order of signed inputs.

Reviewer 2: Clarity nit.

```suggestion
> **Note** In contrast to signing a transaction builder, when signing a built transaction, you must ensure that the order of signatures matches the order of signed inputs.
```
